### PR TITLE
Allow to create a claim or a communication using CODE references

### PR DESCRIPTION
**Pull request by toch**
When the claim mobile app creates a new claim, it doesn't have the UUID of the provider, enterer, and the patient, but it has their CODE. In order to minimize the number of HTTP requests, we made possible to use the CODE as reference when creating a new claim.

**Update 14 Jul 2023**

The initial suggestion was to allow to identify the `patient`, `enterer`, `provider` with a reference using the CODE id rather than the UUID:

```json
    "patient": {
        "reference": "Patient/999000001"
    }
```

As suggested by @delcroip , we're going to use directly the identifier type in the reference as it follows:

```json
{
    "patient":
    {
        "type": "Patient",
        "identifier":
        {
            "value": "CODE_TO_SEARCH",
            "type":
            {
                "coding":
                [
                    {
                        "system": "https://openimis.github.io/openimis_fhir_r4_ig/CodeSystem/openimis-identifiers",
                        "code": "Code"
                    }
                ]
            }
        }
    }
}
```

** Update 22 aug ** same has been done when creating a communication to id the subject

### CODE REVIEW

**Comment by delcroip** [comment]
Hi,

a proper way to do it would be to use the identifier in the reference (_get_type_and_id_from_reference to be updated)
reference = {
   'type'='Patient'
   'identifier':{
         'value':'CODE_TO_SEARCH',
      "[type](http://hl7.org/fhir/R4/datatypes.html#Identifier#Identifier.type)" : {
        "[coding](http://hl7.org/fhir/R4/datatypes.html#CodeableConcept#CodeableConcept.coding)" : [
          {
            "[system](http://hl7.org/fhir/R4/datatypes.html#Coding#Coding.system)" : "https://openimis.github.io/openimis_fhir_r4_ig/CodeSystem/openimis-identifiers",
            "[code](http://hl7.org/fhir/R4/datatypes.html#Coding#Coding.code)" : "Code"
          }
        ]
      },
   }
}



the equivalent in a get serach would be 
Patient?identifier=CODE-TO-SEARCH&identifier.type=https://openimis.github.io/openimis_fhir_r4_ig/CodeSystem/openimis-identifiers|Code

**Comment by delcroip** [comment]
here a fct to manage it 
```
    @classmethod
    def _get_type_and_id_from_reference(cls, reference: Reference) -> Tuple[str, str]:
        """
        Extracts resource type and resource id from FHIR reference.
        first check reference.refrence eg Patient/uuid
        if not present, check type  (eg patient) and identifier value and type
        """
        resource_id, path, code_type = None, None, None
        if reference:
            reference_str = reference.reference
            if isinstance(reference_str, str) and '/' in reference_str:
                path, resource_id = reference_str.rsplit('/', 1)
            elif isinstance(reference.type, str) and reference.identifier is not None\
                 and isinstance(reference.identifier.value, str):
                path = reference.type
                resource_id = reference.identifier.value
                if reference.identifier.type is not None
                    code_type = reference.identifier.type
        if path is None or resource_id is None:
            raise FHIRRequestProcessException(
                [F'Invalid format of reference `{reference}`. '
                 F'Should be string with format <ReferenceType>/<ReferenceId>']
            )
        return path, resource_id, code_type
```

**Comment by toch** [comment]
thanks @delcroip :bow: 

I'll update my PR with this suggestion

**Comment by toch** [comment]
@delcroip I've updated my PR with your suggestion

**Comment by toch** [comment]
As I wasn't able to reproduce it locally (I made sure I was using the last version of `database_ms_sqlserver`), I've reverted the commit to see if we get the same error without my changes ... and it's the case.

The error occurs when testing the contract API (`test_post_should_create_correctly`, and `test_post_patient_subject_reference`) and it is due to an attempt to insert a value in a timestamp field, which is not permitted. 

```
[23000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Cannot insert an explicit value into a timestamp column. Use INSERT with a column list to exclude the timestamp column, or insert a DEFAULT into the timestamp column. (273) (SQLExecDirectW)
```

Does it ring a bell for anyone?

**Comment by toch** [comment]
After several attempts with different approaches, I couldn't reproduce the error locally. Without access to the CI, it's difficult to debug and get more info. I might miss something ofc, some help would be very welcome :+1:.

@delcroip could you have a look please? thanks :bow:

**Comment by toch** [comment]
Thanks to @tdethier we found that the [error occurred](https://github.com/openimis/openimis-be_py/actions/runs/5523256639/jobs/10073927202) in a [recent change of the main repo `openimis-be_py`](https://github.com/openimis/openimis-be_py/pull/163).

So it seems indeed unrelated to the changes suggested here. I'm going to push them again.

**Comment by tdethier** [comment]
Hi @delcroip @dragos-dobre the changes that Patrick suggested were done, but the CI is failing for a reason outside of this PR (if you check all the failing PRs in this module, or the PR in be_py mentioned above, you see the same error message every time). We need this PR to be merged in order to continue working on the mobile apps, could you please check it and merge once you're good with it? Thanks :slightly_smiling_face:

**Comment by toch** [comment]
I've just added two small changes to apply the same search capability by code for medication and activitydefinition

File: api_fhir_r4/converters/claimConverter.py [hunk 1]
```diff
@@ -27,6 +27,8 @@
 
 from api_fhir_r4.utils import TimeUtils, FhirUtils, DbManagerUtils
 
+import logging
+logger = logging.getLogger(__name__)
```
Comment (delcroip) [inline api_fhir_r4/converters/claimConverter.py:31]: To be able to catch all log in the main logger @edarchis is this the right way ?

```suggestion
logger = logging.getLogger('openimis.' + __name__)
```

**Review by delcroip — APPROVED** [review]

**Comment by toch** [comment]
Thanks @delcroip for the fix in the fixture json. For the last one, I know the issue. We need clean up after creation or it complains there is already one communication for that claim. I'll fix that.

**Comment by delcroip** [comment]
I think the mapping done by @dragos-dobre is not optimal, this should take the form of a CGP-collectWith task where we attach a FHIR questionnaire (way more flexible)

Request feedback should create a task (with collectWith extension), 
then the Mobile should get the task and fetch the questionnaire
then it should update the task as done and sent the questionnaireResponse

With that approach we could foresee advanced questionnaire (I can make a demo of android-fhir)
@dborowiecki  this could be use for Grievance too

**Comment by delcroip** [comment]
> Thanks @delcroip for the fix in the fixture json. For the last one, I know the issue. We need clean up after creation or it complains there is already one communication for that claim. I'll fix that.

if it is only that then it is a good news (no regression)

**Comment by toch** [comment]
thanks @delcroip I'll do the test fix/improvement in another PR then.

**Comment by toch** [comment]
@delcroip #146 should do the job regarding the test

**Pull request merged**